src/workerd/io/actor-sqlite.c++

```diff
@@ -169,11 +200,67 @@ void ActorSqlite::onWrite() {
       // rather than after the callback.
       { auto drop = kj::mv(txn); }
 
-      return commitCallback();
+      return commitImpl(kj::mv(precommitAlarmPromise));
     })));
   }
 }
 
+kj::Promise<void> ActorSqlite::requestScheduledAlarm(kj::Maybe<kj::Date> requestedTime) {
+  co_await hooks.scheduleRun(requestedTime);
+  lastConfirmedScheduledAlarm = requestedTime;
+}
+
+ActorSqlite::PrecommitAlarmState ActorSqlite::startPrecommitAlarmScheduling() {
+  PrecommitAlarmState state;
+  state.localAlarmState = metadata.getAlarm();
+  if (pendingCommit == kj::none &&
+      willFireEarlier(state.localAlarmState, lastConfirmedScheduledAlarm)) {
+    // Basically, this is the first scheduling request that commitImpl() would make prior to
+    // commitCallback().  We start the request separately, ahead of calling sqlite functions that
+    // commit to local disk, for correctness in workerd, where alarm scheduling and db commits are
+    // both synchronous.
+    state.schedulingPromise = requestScheduledAlarm(state.localAlarmState);
+  }
+  return kj::mv(state);
+}
+
+kj::Promise<void> ActorSqlite::commitImpl(ActorSqlite::PrecommitAlarmState precommitAlarmState) {
+  // We assume that exceptions thrown during commit will propagate to the caller, such that they
+  // will ensure cancelDeferredAlarmDeletion() is called, if necessary.
+
+  KJ_IF_SOME(pending, pendingCommit) {
+    co_await pending.addBranch();
+    co_return;
+  }
+  auto [promise, fulfiller] = kj::newPromiseAndFulfiller<void>();
+  pendingCommit = promise.fork();
+
+  // Wait for any precommit alarm scheduling work to complete.
+  auto localAlarmState = precommitAlarmState.localAlarmState;
+  KJ_IF_SOME(p, precommitAlarmState.schedulingPromise) {
+    co_await p;
+    localAlarmState = metadata.getAlarm();
+  }
+
+  while (willFireEarlier(localAlarmState, lastConfirmedScheduledAlarm)) {
+    co_await requestScheduledAlarm(localAlarmState);
+    localAlarmState = metadata.getAlarm();
+  }
+
+  auto commitCallbackPromise = commitCallback();
+
+  // Fulfill and clear the pending commit.
+  fulfiller->fulfill();
```

Comment:
I think we actually don't want to fulfill this until the commit callback completes, otherwise we're potentially prematurely unblocking the output gate.

We do still want to set `pendingCommit = none` here so that no _new_ commits merge with this one, but any that have already merged should wait until the commit is actually confirmed.

Reply:
Hmm... what you're proposing makes sense, but I'm a bit puzzled because I have a test case where several invocations of `commitImpl()` are in-flight, and the output gate for all of them does not unblock until the commit promise completes.

https://github.com/cloudflare/workerd/blob/935e9ecdd8ddda26547e49da6b3dc951cfa6db32/src/workerd/io/actor-sqlite-test.c%2B%2B#L640

I suppose this must be because earlier output gate promises block all the later ones?  If so, I'm having trouble thinking of a sequence of events where moving the fulfill() call later results in a difference in behavior, such that I can add a test against a bug.  But I can update the code, so that we're not relying strictly on the output gate behavior for correctness.